### PR TITLE
Balance command

### PR DESCRIPTION
**Pull request by Pennedrose**
## Overview

This PR will.. create a command that automatically balances the robot onto the charging station. 

## Tests Ran

Built code
Haven't tested on robot bc no community room :(

### CODE REVIEW

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        m_drive = drive;
+    }
+
+    /*
+     * Determines the values of the vectors of pitch, yaw, and roll. 
+     */
+    public void initialize()  {
+
+        endPoint = 0;
+        startAngle = Functions.calculateHypotenuse(m_drive.getAngleHeading(), Functions.calculateHypotenuse(m_drive.getPigeon().getPitch(), m_drive.getPigeon().getRoll()));
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/BalanceCommand.java:25]: I'm not sure how you got this equation but I'm not sure it's right

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    /**
+     * Makes the robot wiggle around PID like until charge station is balanced. 
+     * ySpeed is zero for now, will add yaw at later pull request.
+     *  
+     */
+    public void execute()  {
+    
+      m_drive.m_balanceController.setTolerance(0.1);
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/BalanceCommand.java:43]: tolerance can be a constant

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+      if(mode)   {
+        m_drive.driveHeading(m_drive.m_balanceController.calculate(endPoint, startAngle), 0, 0, true);
+      }
+      else   {
+        m_drive.driveHeading(m_drive.m_balanceController.calculate(endPoint, startAngle), 0, 90, true);
+      }
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/commands/BalanceCommand.java:51]: endpoint and start angle are in the wrong places the set point should be the second paramater and the messuement the first. Additionaly start angle (the messurement) is never updated

**Comment by PandaNinja10** [comment]
Something of note, shouldn't we all ways use roll as we want the robot to have centered mass. also we should lock the robot to the closest increment of 90 or 180 degrees rather than only 0 or 90 if we are going to use pitch.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 4]
```diff
@@ -54,6 +59,8 @@ public class Drivetrain extends SubsystemBase {
 
   private double m_headingPIDOutput = 0;
 
+  
+
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:63]: too many random space for my taste

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 5]
```diff
@@ -266,6 +281,11 @@ public Rotation2d getRotation2d() {
   public double getAngleHeading() {
     return MathUtil.angleModulus(m_pigeon.getRotation2d().getRadians());
   }
+
+  public Pigeon2 getPigeon()
+  {
+    return m_pigeon;
+  }
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:288]: there is no reason at the momment to access the Peigeon directly, make getter methods for what you need from it

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 6]
```diff
@@ -89,13 +97,19 @@ public class Drivetrain extends SubsystemBase {
...
+
   /**
    * Creates a new Swerve Style Drivetrain.
    * @param drivetrainTab the shuffleboard tab to display drivetrain data on
    * @param swerveModulesTab the shuffleboard tab to display module data on
    */
   public Drivetrain(ShuffleboardTab drivetrainTab, ShuffleboardTab swerveModulesTab) {
 
+    m_config.configureControls();
+
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:112]: don't make a new driver controller or game controller in drivetrain, you don't even use it for anything

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    private boolean mode;
+ 
+    public BalanceCommand(Drivetrain drive) {
+        m_drive = drive;
+    }
+
+    /*
+     * Determines the values of the vectors of pitch, yaw, and roll. 
+     */
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/commands/BalanceCommand.java:21]: this is more an in code commnet than a comment for the whole method you can probobly leave the initalize method un-java-documented.

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import frc.robot.subsystems.Drivetrain;
+import frc.robot.util.Functions;
+
+public class BalanceCommand extends CommandBase {
+
+    private Drivetrain m_drive;
+
+    private double startAngle, endPoint;
+
+    private boolean mode;
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/commands/BalanceCommand.java:13]: mode is to vauge of a variable name and variable should have "m_" prefix

File: src/main/java/frc/robot/controls/Operator.java [hunk 9]
```diff
@@ -24,7 +26,7 @@ public Operator(FourBarArm arm, Intake intake){
     m_intake = intake;
   }
 
-  public void configureControls() {
+  public void configureControls(Drivetrain drive) {
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/controls/Operator.java:29]: have the drivetrain be input in the contructor

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 10]
```diff
@@ -125,8 +139,9 @@ public void periodic() {
     updateOdometry();
     
     m_fieldDisplay.setRobotPose(getPose());
-  }
-  
+     
+    }
+
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:144]: remove misindentation

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        m_inverted = 1;
+
+        m_pid.setSetpoint(0);
+        if(Math.abs(Math.PI/2 - m_drive.getAngleHeading() % Math.PI) > Math.PI/4)  { //Determines whether to use roll or pitch
+            m_usePitch = false;
+        }
+        else  {
+            m_usePitch = true;
+        }
+        if(m_drive.getAngleHeading() > Math.PI)
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/commands/BalanceCommand.java:38]: The getHeaddingAngle method is modulus and will only output values from -PI to PI so there is no situation in wich a value greater than PI will be output. I think you will have to change it to value<0

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public void execute() {
+        m_output = MathUtil.clamp(m_pid.calculate(m_currentAngle, DriveConstants.kBalanceSetpoint), -DriveConstants.kBalanceMaxOutput, DriveConstants.kBalanceMaxOutput);
+        if(m_usePitch) {
+            m_drive.driveHeading(m_output, 0, m_inverted*Math.PI/2, true);
+        }
+        else {
+            m_drive.driveHeading(m_output, 0, 0, true);
```
Comment (PandaNinja10) [inline src/main/java/frc/robot/commands/BalanceCommand.java:51]: heading optimization is acounted for in pitch but not yaw

File: build.gradle [hunk 13]
```diff
@@ -1,6 +1,6 @@
 plugins {
     id "java"
-    id "edu.wpi.first.GradleRIO" version "2023.3.2"
+    id "edu.wpi.first.GradleRIO" version "2023.4.1"
```
Comment (AnthonyFurman) [inline build.gradle:3]: Make this on a separate PR @PandaNinja10

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        if(m_drive.getAngleHeading() < 0)
+        {
+            m_inverted = -1;
+        }
+    }
+
+    @Override
+    public void execute() {
+        new PrintCommand("Banana");
+        m_output = MathUtil.clamp(m_pid.calculate(m_currentAngle, DriveConstants.kBalanceSetpoint), -DriveConstants.kBalanceMaxOutput, DriveConstants.kBalanceMaxOutput);
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/BalanceCommand.java:48]: m_pid.calculate() can take just one parameter, the process variable (the current angle) and you don't need to give the setpoint each time. The setpoint of zero doesn't meed to be a constant I think.

File: src/main/java/frc/robot/commands/BalanceCommand.java (hunk 12 above)
Comment (AnthonyFurman) [reply src/main/java/frc/robot/commands/BalanceCommand.java:53]: Is this optimization necessary? The driving is field relative, so positive will always be forward, even if the robot is facing backward.
One thing you guys will have to figure out is if you go on the charge station from the other side, but that's something different and only for auto as the driver in teleop would be able to get on the station.

File: src/main/java/frc/robot/commands/BalanceCommand.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public boolean isFinished() {
+        return m_pid.atSetpoint();
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+        m_drive.stop();
+        //TODO: Make the wheels into X to prevent rolling off
+    }
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/BalanceCommand.java:66]: I would just schedule it as a sequence command group if you want that to happen, not have it as part of this command. So each command does 1 thing.

File: src/main/java/frc/robot/constants/swerve/DriveConstants.java [hunk 16]
```diff
@@ -18,7 +18,7 @@ public class DriveConstants {
   public static double kDriveGearRatio = 6.75;
   public static double kSteerGearRatio = 21.43;
 
-  public static double kMaxSpeed = (FalconConstants.kMaxRpm / 60.0) * kDriveGearRatio * kWheelRadius * 2 * Math.PI;
+  public static double kMaxSpeed = (FalconConstants.kMaxRpm / 60.0) / kDriveGearRatio * kWheelRadius * 2 * Math.PI;
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/constants/swerve/DriveConstants.java:21]: Separate PR, approve immediately @PandaNinja10 @richtan

File: src/main/java/frc/robot/controls/ManualController.java [hunk 17]
```diff
@@ -28,5 +32,7 @@ public void configureControls() {
     m_manual.get(Button.Y).onTrue(new InstantCommand(() -> m_arm.setMotorPower(0.05)));
     m_manual.get(Button.X).onTrue(new InstantCommand(() -> m_arm.setMotorPower(-0.05)));
     m_manual.get(Button.B).onTrue(new InstantCommand(() -> m_arm.setMotorPower(0)));
+
+    m_manual.get(Button.RB).onTrue(new BalanceCommand(m_drive));
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/controls/ManualController.java:36]: Shouldn't this be in test controls?

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 18]
```diff
@@ -65,6 +68,7 @@ public class Drivetrain extends SubsystemBase {
   private final PIDController m_xController = new PIDController(0.1, 0, 0);
   private final PIDController m_yController = new PIDController(0.1, 0, 0);
   private final PIDController m_rotationController = new PIDController(DriveConstants.kHeadingP, DriveConstants.kHeadingI, DriveConstants.kHeadingD);
+  public final PIDController m_balanceController = new PIDController(DriveConstants.kBalanceP, DriveConstants.kBalanceI, DriveConstants.kBalanceD);
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:71]: This PID controller isn't used as far as I can tell.

File: src/main/java/frc/robot/controls/Operator.java [hunk 19]
```diff
@@ -1,26 +1,43 @@
...
+    // elevator controls
+    operator.get(Button.Y).onTrue(new ExtendToPosition(m_arm, ArmConstants.kTopPosition));
+    operator.get(Button.X).onTrue(new ExtendToPosition(m_arm, ArmConstants.kMiddlePosition));
+    operator.get(Button.A).onTrue(new ExtendToPosition(m_arm, ArmConstants.kIntakePosition));
+    operator.get(Button.B).onTrue(new ExtendToPosition(m_arm, ArmConstants.kShelfPosition));
 
+    // intake controls
+    operator.get(DPad.DOWN).onTrue(new InstantCommand(() -> m_intake.intake(IntakeConstants.kIntakeSpeed), m_intake));
+    operator.get(DPad.UP).onTrue(new InstantCommand(() -> m_intake.intake(IntakeConstants.kOuttakeSpeed),m_intake));
+    operator.get(DPad.LEFT).onTrue(new InstantCommand(() -> m_intake.stop(), m_intake));
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/controls/Operator.java:40]: All these changes are not related to the balance command, remove them?

File: src/main/java/frc/robot/commands/DefaultDriveCommand.java [hunk 20]
```diff
@@ -27,8 +28,9 @@ public void execute() {
     double xSpeed = m_driver.getForwardTranslation();
     double ySpeed = m_driver.getSideTranslation();
     double rot = m_driver.getRotation();
+    double slowFactor = m_driver.isSlowSpeed().getAsBoolean() ? 0.5 : 1;
 
-    m_drive.drive(xSpeed, ySpeed, rot, true);
+    m_drive.drive(xSpeed * slowFactor, ySpeed * slowFactor, rot, true);//m_driver.isSlowSpeed().getAsBoolean()
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/DefaultDriveCommand.java:33]: All this slow speed stuff should be in a separate PR so it can be approved quickly

File: src/main/java/frc/robot/commands/DefaultDriveCommand.java [hunk 21]
```diff
@@ -10,9 +11,9 @@
 public class DefaultDriveCommand extends CommandBase {
 
   private final Drivetrain m_drive;
-  private final BaseDriverConfig m_driver;
+  private final GameControllerDriverConfig m_driver;
```
Comment (AnthonyFurman) [inline src/main/java/frc/robot/commands/DefaultDriveCommand.java:15]: You should still be using BaseDriverConfig
